feat: Add Chat Context Awareness

Davixk opened the pull request:
# Changelog Entry

### Description

Implements automatic Chat Context injection that adds recent or pinned chat history to conversation context.
I believe this feature will enhance the user experience by personalizing model context. Find an example in the screenshot below.

### Added

- Backend config: `ENABLE_CHAT_CONTEXT`, `CHAT_CONTEXT_MODE`, `CHAT_CONTEXT_TOP_K` in `config.py`
- Middleware handler `chat_context_handler()` in `utils/middleware.py` supporting:
  - Recent mode: retrieves last N chats (skip=1 to exclude current)
  - Pinned mode: retrieves user-pinned chats (early return if none exist)
  - Structured extraction: first_user_message + last_messages{user, assistant}
  - Conditional prompts: encouraging for pinned, discouraging for recent
  - JSON-in-XML injection format
- Admin UI in `General.svelte`: enable toggle, mode dropdown, top K input
- Feature flag in `Chat.svelte`: checks `$config.chat_context.enable`
- Config exposure via `/api/config` endpoint

### Changed

- `AdminConfig` model now includes chat context fields
- `get_admin_config()` and `update_admin_config()` handle chat context settings
- Middleware pipeline calls `chat_context_handler()` when feature enabled

### Additional Information

**Design decisions:**
- Skip=1 in recent mode prevents self-reference
- No skip in pinned mode
- Early return for empty pinned chats
- Multimodal content support extracts text from arrays

**Chat format injected:**
\`\`\`json
[
  {
    "title": "Chat Title",
    "created_at": "2024-01-15T10:30:00-0500",
    "updated_at": "2024-01-15T11:45:00-0500",
    "first_user_message": "...",
    "last_messages": {
      "user": "...",
      "assistant": "..."
    }
  }
]
\`\`\`

**Testing:** Manual testing of enable/disable, mode switching, top K validation, pinned empty/full, recent mode, config persistence, feature flag, JSON structure, multimodal handling.
### Screenshots or Videos

<img width="1014" height="856" alt="image" src="https://github.com/user-attachments/assets/2abf0756-59c7-4728-9914-9581f5d10082" />
<img width="1012" height="856" alt="image" src="https://github.com/user-attachments/assets/eaf55000-51fd-4cf9-8e09-60ebb12ec6fe" />

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

Davixk committed:
feat(chat-context): add configurable chat context injection

Introduces admin-configurable chat context features, enabling automatic injection of recent or pinned chats into conversation context to improve continuity across sessions.

Adds UI controls for enabling, selecting mode, and setting the number of chats included. Lays groundwork for future "relevant" AI-based selection.

Davixk commented:
thank you @EntropyYue !

let me know if you have any feature suggestions.

my goal is to make the models as context aware as possible - while keeping token usage efficient and agnostic.

Davixk commented:
thank you. setting to draft until I can work on it.

silentoplayz commented:
Unrelated, but why is your entire UI font in lowercase? 😄

Edit: `URL` is still uppercase! LOL

silentoplayz commented:
Also, I encourage you to keep going with this PR. This is quite nice if you asked me. 
<img width="2300" height="435" alt="image" src="https://github.com/user-attachments/assets/2e65064f-e50c-4d0a-80f0-661c5ed27553" />

Davixk commented:
@rgaricano there is a solid argument to be made about that! and it is something I thought about.

here's why I think it **should** indeed be a core feature:
1. there is a **high enough advantage** to gain, even just by injecting core info from the last 2/3 chats. it doesn't necessarily need embedding and similarity retrieval to become a useful feature
2. this can **affect future plans** for Open WebUI: as you said yourself, some great future additions might come in the form of chat embedding and retrieval, and I have been thinking about **AI chat summaries**, which could be an asynchronous task that the WebUI runs for each chat, which would be best stored in a new dedicated field, and might need db changes. 
3. usefulness is **proven by ChatGPT**. it has implemented a very similar feature in May 2025, where the last few most recent chats are injected into the prompt, and I believe it works well enough

you can find more info in my issue #13595 from May 6th, where Tim seemed to agree with me.

rgaricano commented:
yes, but I can do that without need of an auto inyect, I just add to context the chats that I want.

https://github.com/user-attachments/assets/009cd8fc-5ebc-4831-9643-a22183cbc416

Davixk commented:
@rgaricano i get what you're saying. and sure, manually adding context works - but that's a power user workflow.

you can also manually add memories to every chat by typing them in, or copy-paste messages from other chats every time. that doesn't mean we shouldn't automate it.

the goal here is to make context awareness **automatic and seamless** for everyone, not just people who proactively drag & drop. most users won't even think to do that, but they'll absolutely benefit from the model remembering their last conversation without lifting a finger.

manually attaching chats is a great feature, but it doesn't really conflict with or detract from this PR - both are useful in their own ways. this just raises the baseline experience while you can still have full manual control when you want it.

Davixk committed:
refactor(admin-settings): improve chat context UX with tooltips

Enhances the chat context settings UI by integrating tooltips directly into labels for better clarity and accessibility. Updates layout for improved alignment and consistency, and clarifies option names. Aims to make configuration options more intuitive for administrators.

Davixk committed:
refactor(middleware): centralize message content cleaning

Moves logic for stripping reasoning blocks and images from message content into a dedicated utility function, ensuring consistent and maintainable content sanitization across the codebase. Improves clarity and reduces code duplication in message processing.

tjbck commented:
This should be implemented as a Filter.
